File: src/Java/algorithms/rcanddp/JumpGame.java
package algorithms.rcanddp; 

import java.util.*; 

public class JumpGame {
    /* Function: Given an array shows on every position how far can jump, 
     *     check if it can reach the end. 
     * Solution: Maintain two pointers. One shows how far it can reach, 
     *     another shows jump from where. If reach a position that cannot reach
     *     before, return false; 
     */ 
    
    public boolean canJump(int[] A) {
        if (A == null || A.length <= 1) {
            return true; 
        }
        
        int i = 0, far = 0; 
        while (far >= i) {
            far = Math.max(far, i + A[i++]); 
            if (far >= A.length - 1) {
                return true; 
            }
        }
        
        return false; 
    }
}

File: src/CPP/src/leetcode/JumpGame.cpp
/*
 * [Source] https://leetcode.com/problems/jump-game/
 * [Difficulty]: Medium
 * [Tag]: Array
 * [Tag]: Greedy
 */

#include <iostream>

using namespace std;

// [Solution]: Use a pointer to track the farest position. If reach that position but cannot move forward, return false.
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/rcanddp/JumpGame.java
 */

/* Java solution
public class Solution {
    public boolean canJump(int[] A) {
        if(A.length<=1) return true;
        int far=0;
        int i=0;
        while(far>=i){
            if(i+A[i]>far) far=i+A[i];
            i++;
            if(far>=A.length-1) return true;
        }
        return false;
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
